core/dbt/adapters/base/connections.py

```diff
@@ -159,6 +162,131 @@ def set_connection_name(self, name: Optional[str] = None) -> Connection:
+            debug, and call warning on unexpected errors or when all
+            retry attempts have been exhausted.
+        :param int retry_limit: How many times to retry the call to
+            acquire_handle. If this limit is exceeded before a successful
+            call, a FailedToConnectException will be raised.
+        :param bool retry_all: Retry on all errors, regardless if they have
+            handlers defined in exception_handlers or not.
+        :param int timeout: How many seconds to wait between attempts to
+            acquire_handle.
+        :param dict exception_handlers: A mapping of exception classes
+            that should be retried to functions to run upon catching them.
+            The values may be set to None if no handling is required before
+            retrying.
+        :param Credentials credentials: A subclass of Credentials that implements
+            acquire_handle to override credentials in connection.
+        :param kwargs: Optional additional keyword arguments to be passed to
+            Connection.credentials.acquire_handle.
+        :raises dbt.exceptions.FailedToConnectException: Upon exhausting all retry
+            attempts without successfully acquiring a handle, or upon the first
+            non-expected Exception if retry_all is set to False.
+        :return: The given connection with its appropriate state and handle attributes
+            set depending on whether we successfully acquired a handle or not.
+        """
+        attempt = 1
+        latest_exc = None
+        expected_exceptions = exception_handlers.keys() if exception_handlers else ()
+        exception_handlers = exception_handlers if exception_handlers else {}
+        max_attempts = 1 + retry_limit
+        acquire_handle = (
+            credentials.acquire_handle
+            if credentials is not None
+            else connection.credentials.acquire_handle
+        )
+
+        while attempt <= max_attempts:
+            try:
+                handle = acquire_handle(**kwargs)
+
+            except Exception as e:
+                latest_exc = e
+
+                is_expected = any(issubclass(type(e), exc) for exc in expected_exceptions)
+                if is_expected:
+                    logger.debug(
+                        "Got an error when attempting to open a "
+                        "{conn_type} connection. Retrying due to "
+                        "either retry configuration set to true."
+                        "This was attempt number: {attempt} of "
+                        "{max_attempts}. "
+                        "Retrying in {timeout} "
+                        "seconds. Error: '{error}'".format(
+                            conn_type=cls.TYPE,
+                            attempt=attempt,
+                            max_attempts=max_attempts,
+                            timeout=timeout,
+                            error=e,
+                        )
```

Comment:
:+1: I also like f-strings. That being said, personally I would have gone with %s-style formatting (or `{}`) to allow `logging` to defer message formatting as the [Python docs describe](https://docs.python.org/3/howto/logging.html#optimization). However, this was not possible as we are using [`AdapterLogger`](https://github.com/dbt-labs/dbt-core/blob/43e3fc22c4eae4d3d901faba05e33c40f1f1dc5a/core/dbt/events/adapter_endpoint.py#L12=) instead of plain `logging`.

Reply:
Oh you can still do that with the AdapterLogger. [Here's an example from the postgres adapter.](https://github.com/dbt-labs/dbt-core/blob/43e3fc22c4eae4d3d901faba05e33c40f1f1dc5a/plugins/postgres/dbt/adapters/postgres/connections.py#L78).